Validate user messages on chat completions endpoint

mdevino opened the pull request:
Closes #382.

mdevino commented:
> For this particular endpoint currently, where `/content` detectors are invokable on chat completions input or output, we would expect text content to be provided for the last message, with any of the roles that this endpoint will work for. The role validation is done here https://github.com/evaline-ju/fms-guardrails-orchestrator/blob/6bc045f0768bc9f9688c47a2dc5f7605a01b4d96/src/orchestrator/handlers/chat_completions_detection/unary.rs#L121 on input, so I’m curious why we want to check at the deserialization separately, if the deserialization is not where the error occurs?

Thanks for catching that. I'm not super familiar with the codebase for this endpoint yet, so I wasn't sure that was the best place to put the validation.
I've updated the code to validate it where you suggested. The one thing I noticed is that, when made the change requested, it doesn't fix the issue if no input detectors are provided (it does fix it if an input detector is provided, though).

Request
```json
{
		"messages": [
				{
					"role": "user",
					"name": "string"
				}
			],
			"model": "my_model_id",
			"n": 1,
			"temperature": 1,
			"top_p": 1,
			"user": "user-1234",
			"detectors": {}
}
```

Response
```json
{
	"code": 422,
	"details": "chat completion request failed for `my_model_id`: list index out of range"
}
```

declark1 commented on src/clients/openai.rs:
```diff
@@ -219,7 +219,17 @@ impl TryFrom<Map<String, Value>> for ChatCompletionsRequest {
         }
         let messages = if let Some(messages) = value.get("messages") {
             Vec::<Message>::deserialize(messages)
-                .map_err(|_| ValidationError::Invalid("error deserializing `messages`".into()))
+                .map(|messages| {
+                    for message in &messages {
+                        if message.role == Role::User && message.content.is_none() {
+                            return Err(ValidationError::Invalid(
+                                "`content` is mandatory for user messages".into(),
+                            ));
+                        }
+                    }
+                    Ok(messages)
+                })
+                .map_err(|_| ValidationError::Invalid("error deserializing `messages`".into()))?
         } else {
             Err(ValidationError::Required("messages".into()))
         }?;
```
Not sure if this is finished yet, but a more ergonomic way to write this would be below. It's good to use `any` / `all` iterator methods for stuff like this.

```rust
let messages = if let Some(messages) = value.get("messages") {
    Vec::<Message>::deserialize(messages)
        .map_err(|_| ValidationError::Invalid("error deserializing `messages`".into()))
} else {
    Err(ValidationError::Required("messages".into()))
}?;
if messages
    .iter()
    .any(|message| message.role == Role::User && message.content.is_none())
{
    return Err(ValidationError::Invalid(
        "`content` is required for user messages".into(),
    ));
};
```

Btw, this just validates that `Content` is provided, but the inner value `Content::Text(text)` could still be an empty `String`. Do we want to validate that it isn't empty?

mdevino replied:
It's not the final code yet. I just pushed what I have before going on vacations, so someone could pick up form there if needed. I like the `any()` suggestion!

mdevino commented:
I've updated the code with the following changes:
1. Rebased to latest main.
2. Updated the validation condition: now a 422 will be returned whenever the last message has no content, regardless of role, and input detectors are provided.
3. Added a few integration tests for 1) no detectors scenarios (passthrough scenarios must pass) and 2) make sure the validation is working.

What I did **not** do:
1. Added a warning when output detectors are provided and all choices have empty content - this is being more complex than expected and will be addressed on a different PR).
2. Added deep check for empty content - At this point in time, I'm just checking if `Message.content` is `None`. For a further validation (such as empty strings), I'd create a method on an `impl Message` block, similar to this:


```rust
impl Message {
    pub fn is_text_content_empty(&self) -> bool {
        match &self.content {
            Some(content) => match content {
                Content::Text(string) => string.is_empty(),
                Content::Array(content_parts) => {
                    content_parts.is_empty()
                        && content_parts.iter().all(|content_part| {
                            content_part.text.is_none()
                                && content_part.text.as_ref().unwrap().is_empty()
                        })
                }
            },
            None => true,
        }
    }
}
```

Let me know if we'd like to go down this route, as content_parts validation

mdevino commented on src/clients/openai.rs:
```diff
@@ -259,6 +259,12 @@ impl ChatCompletionsRequest {
                 "`messages` must not be empty".into(),
             ));
         }
+        if !self.detectors.input.is_empty() && self.messages.last().unwrap().content.is_none() {
```
There's no link. How would you like to go about that?
I can only think about adding a todo comment, but that doesn't really mean linking.

mdevino commented on tests/chat_completions_detection.rs:
I could. What would be a real scenario in this case, a tool call message being the last?

mdevino commented on tests/chat_completions_detection.rs:
```diff
@@ -1185,5 +1186,139 @@ async fn orchestrator_validation_error() -> Result<(), anyhow::Error> {
+            ..Default::default()
+        },
+        Message {
+            content: Some(Content::Array(vec![
```
Content as array is supported on passthrough, but if the last message's content is an array, it is being sent to the detector as an empty string. I've updated the validation not to allow arrays and will raise a story for us to discuss how we want to support detection on messages with content as an array.

mdevino commented on src/clients/openai.rs:
```diff
@@ -259,6 +259,26 @@ impl ChatCompletionsRequest {
             ));
         }
+
+        // Content of type Array is not supported yet
+        // Adding this validation separately as we do plan to support arrays of string in the future
+        if !self.detectors.input.is_empty() {
+            if let Some(Content::Array(_)) = self.messages.last().unwrap().content {
+                return Err(ValidationError::Invalid(
+                    "Detection on array is not supported".into(),
+                ));
+            }
+        }
+
+        // As text_content detections only run on last message at the moment, only the last
+        // message is being validated.
+        if !self.detectors.input.is_empty() && self.messages.last().unwrap().is_text_content_empty()
+        {
+            return Err(ValidationError::Invalid(
+                "if input detectors are provided, `content` must not be empty on last message"
+                    .into(),
+            ));
+        }
```
I've left a comment about thisbon line 264.

mdevino replied:
Done.
